Prometheus: properly check for proxy mode

GiedriusS opened the pull request:
The original check has been added ~3 years ago and since then a new variable has been added that says whether the mode is browser or proxied.

Start using the new variable in this function so that proxy mode could be properly detected. It's important to us to get these headers and this check is currently broken for HTTP/HTTPS URLs.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=54823) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=54823) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=54823) it.</sub>

GiedriusS commented:
@itsmylife tried fixing the tests but they are now failing with:

```
    TypeError: Cannot read property 'fetch' of undefined                                             
```

Seems like mocking doesn't work properly with `jest` for some reason after these changes. If I import `getBackendSrv` and `setBackendSrv` from `@grafana/runtime` and use them for mocking then I can get past that error. But, after that it seems like `toDataQueryResponse` is somehow broken - it sets an error field even though no error has occurred. It checks for `  if ((res as FetchResponse).status && (res as FetchResponse).status !== 200) {` but status is `success`. I didn't have time after that to see what is happening. Please help me out.

itsmylife commented on public/app/plugins/datasource/prometheus/datasource.test.ts:
```diff
@@ -1783,6 +1783,7 @@ describe('PrometheusDatasource for POST', () => {
     directUrl: 'direct',
     user: 'test',
     password: 'mupp',
+    access: 'proxy',
```
```suggestion
```

itsmylife commented on public/app/plugins/datasource/prometheus/datasource.test.ts:
```diff
@@ -1861,9 +1863,9 @@ describe('PrometheusDatasource for POST', () => {
-      expect(httpOptions.headers['X-Dashboard-Id']).toBe(undefined);
-      expect(httpOptions.headers['X-Panel-Id']).toBe(undefined);
-      expect(httpOptions.headers['X-Dashboard-UID']).toBe(undefined);
+      expect(httpOptions.headers['X-Dashboard-Id']).toBe(1);
+      expect(httpOptions.headers['X-Panel-Id']).toBe(2);
+      expect(httpOptions.headers['X-Dashboard-UID']).toBe('WFlOM-jM1');
```
if you would add the following to this `describe` section all tests will be passed.  Also, you have to remove the `access: 'proxy:'` from above. See my suggestions.
```ts
const instanceSettings = {
      url: 'proxied',
      directUrl: 'direct',
      user: 'test',
      password: 'mupp',
      access: 'proxy',
      jsonData: { httpMethod: 'POST' },
    } as unknown as DataSourceInstanceSettings<PromOptions>;

    let ds: PrometheusDatasource;
    beforeEach(() => {
      ds = new PrometheusDatasource(instanceSettings, templateSrvStub as any, timeSrvStub as any);
    });
```

itsmylife reviewed (commented):
Suggestions to make tests pass

GiedriusS committed:
Prometheus: fix proxy check

The original check has been added ~3 years ago and since then a new
variable has been added that says whether the mode is browser or
proxied.

Start using the new variable in this function so that proxy mode could
be properly detected. It's important to us to get these headers and this
check is currently broken for HTTP/HTTPS URLs.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

GiedriusS commented:
@itsmylife thanks for your help, I finally managed to fix tests. Please take a look again when you have some time <3
